### ISSUE TITLE
Is it possible to use the same AppSync api across multiple serverless projects?

### ISSUE THREAD

**Issue by EwanValentine** [sid88in/serverless-appsync-plugin#194]
I have several microservices, all of which exposed via AppSync. At the moment, I have one AppSync project, and when I deploy a service, it extracts the graphql config, uploads it to S3. Then the AppSync project pulls down the files and rebuilds. But it's a bit convoluted, and difficult to run locally. So I wondered if it was possible to 're-use' the same AppSync api, perhaps by giving it the same api name in the config?

**Comment by EwanValentine** [comment]
@sid88in any thoughts on this?

**Comment by sid88in** [comment]
not sure. did you try this out?

**Comment by EwanValentine** [comment]
@sid88in we did! Unfortunately it created another AppSync API with a different unique key. 

I wonder if it would be possible, internally to provide a 'key', or maybe an ARN. If an AppSync key or Arn is provided in the custom config. It adds the config to the existing AppSync API instead?

**Comment by EwanValentine** [comment]
Willing to help out if you think that this is doable or possible

**Comment by bboure** [comment]
@EwanValentine I'd like to make sure I understood your issue.

So, you have several microservices all exposed through the same graphql.
Are all microservices in different stacks/serverless.yml? 
If so, what you want it to deploy all microservices individually when updated and update only the part of graphql that corresponds to it? (probably a query or mutation field)

**Comment by EwanValentine** [comment]
@bboure they're in different stacks/serverless.yml files, yes. And ideally yes, that might be a crazy idea. We're doing it manually at the moment. So when we deploy a 'graphql service', we upload a GraphQL schema, a yaml config file with data sources, and any `vtl` templates into an S3 bucket. Then that kicks off a Codebuild pipelines to merge them all into one AppSync project. 

It's a bit convoluted, and means it's tricky to run things in isolation. So I was doing some research to see if there was a better way, basically. Hope this helps!

**Comment by bboure** [comment]
@EwanValentine I assume that you are using lambdas for your microservices?

What about this:

Have your microservices only deploy a lambda with your code. No Appsync involved here.
In your AppSync stack, your datasources call a **proxy** lambda whose only job is to call each individual microservice lambda with the received payload and return its result.

That way, each stack is independent. Your proxy lambda will likely never change and remain as it is (unless you need to add a new microservice). You just need to focus on deploying your microservices lambdas. 

Of course this has a couple of drawbacks since each microservices will involve calling 2 lambdas each time: This will add a bit of latency and, of course, cost.
To reduce that, your proxy lambda could be just one same lambda that receives as input the microservice lambda name that has to be called. Doing that should at least reuse it and keep it warm.

Example:
- your microservice stack "sendEmail" has a lambda called `sendEmailLambda` with your code.
- your appsync stack has a mutation "sendEmail" with a datasource that calls a `proxyLambda`. 
- your `sendEmail` template adds the microservice name to the payload `payload: {"microservice: "sendEmailLambda", microservicePayload: $utils.toJson($context)}`
- your lambda `proxyLambda` calls the lambda received in the payload: `microservice` and returns its result directly.

Once everything is in place, when your `sendEmail` service changes, you just need to update the `sendEmailLambda` function. Your appsync stack will not change. 

Would that be suitable for you?

**Comment by kkesley** [comment]
I'm currently having this problem as well.

My project looks like this:

```
/api
   /some-api-1
      api-1-function-1 (every function is a lambda)
      api-1-function-2
      serverless.yml
   /some-api-2
      api-2-function-1
      api-1-function-2
      serverless.yml
serverless.yml
```

I used that for API Gateway and now, I want to change it to appsync.

My problem is that I used to make an API Gateway Resource in the root serverless.yml and refer it in the child serverless.yml. 
**Note: I need to deploy root serverless.yml first, then the children**

it'd be nice to reuse the graphql api in the child serverless.yml

**Comment by bboure** [comment]
@kkesley @EwanValentine 

I had a second look at this, I think that what you need is to be able to manually override [this parameter](https://github.com/sid88in/serverless-appsync-plugin/blob/master/index.js#L600), and specify an AppSync arn from an external stack
In addition to that, you need to be able to define the dataSources in a child stack, without deploying a new AppSync API.

Now, that sounds like something possible, however, there is much more involved here.
For example, the GraphQl schema.  Ideally, the schema part for your child stacks should be defined there too, with the data source. But that means that whenever you want to deploy a child stack, you will need to pull the schema form the parent and siblings' stacks and merge them, in order to push the full schema (schema in AppSync is kinda "atomic").
This does not sound easy and prone to issues: deploying a microservice could affect others. 
I am not sure how we could do that easily.

Any suggestions?

**Comment by kkesley** [comment]
@bboure regarding the schema, would it be much simpler if we just manage it in the parent stack? in my opinion, this is appsync's limitation rather than the library.

**Comment by bboure** [comment]
@kkesley I also see it as an AppSync limitation rather than the plugin.

If you are going to manage the schema in the parent stack, then what I can suggest you instead and would currently be supported is the following: 
- Have all the AppySinc config, the schema and all the data sources in your parent stack.
- Have your lambda functions in the "child" stacks.
- In parent stack, reference your Lambda functions from the child stacks with their ARN. 

To make this easy, you can take advantage of the [Output](https://serverless.com/framework/docs/providers/aws/guide/variables#reference-cloudformation-outputs) feature from Cloudformation/Serverless.

This way, you only need to touch your AppSync stack if the structure our your api (schema) changes.
You microservices will then only take care of deploying the lambdas when your code logic changes. 

How does that sound?

**Comment by kkesley** [comment]
@sid88in that's one way to do it. However, it will clutter the parent stack with `data sources` and `mapping templates`.. 

Furthermore, parent stack would be dependent on child stacks. which means, if you want to deploy it again in a new environment, you would deploy child stacks first and then the parent stack. In my opinion, it's less intuitive.

**Comment by EwanValentine** [comment]
Sorry for the late response! Just read back through all of this. I agree it's an AppSync limitation, probably well out of scope for this particular project. Thanks for looking into it though, I'll take on board some of the other suggestions above

**Comment by EwanValentine** [comment]
Feel free to close, unless you think there's value in keeping this discussion on-going

**Comment by andrejunges** [comment]
any updates on this? facing this issue currently.. seems like we still don't have a "perfect"/recommended solution.

**Comment by bboure** [comment]
@andrejunges it really depends on your use case. 
can you elaborate?

**Comment by andrejunges** [comment]
sure, currently we have multiple services exposing it's own Appsync API...
But we'd like to have them all together, merge these individual schemas into one centralized appsync api.

it seems `apiId` (which was deprecated) could help doing that, but based on the replies here it wouldn't  merge the schemas defined in each service.

Not sure what's the best solution for now, but our ideal solution would be to define the schema, mapping templates and datasources in each service, but deploy them all together using the same AppSync API.

**Comment by bboure** [comment]
@andrejunges What you probably want is Schema federation
This is currently not supported by AppSync, but there is a request for it: https://github.com/aws/aws-appsync-community/issues/35

However someone suggested using this: https://github.com/0xR/graphql-transform-federation

I have not tried it myself but that might be something you can explore? 

Since this seems like something of interest, I will re-open this issue.

**Comment by andrejunges** [comment]
@bboure thanks again for the reply.
I saw that on my researches and I agree that would be the ideal solution - though not sure how well it'd work and was trying to get some feedback first.

when I read about the `apiId` it seemed like a better solution until appsync implements something like apollo federation or at least provide support for it - https://github.com/serverless-components/aws-app-sync#create-or-reuse-apis.
Is it accurate apiId do not merge the schemas? or what was the reason this library deprecated it?

**Comment by SoimanAndrei** [comment]
Any news about supporting [https://github.com/serverless-components/aws-app-sync#create-or-reuse-apis](https://github.com/serverless-components/aws-app-sync#create-or-reuse-apis) ?

**Comment by bboure** [comment]
@SoimanAndrei  Please see this thread: https://github.com/sid88in/serverless-appsync-plugin/issues/300
I gave some clues on how this *might* be achieved. I have not tried it though.

**Comment by tomaszdudek7** [comment]
@kkesley I just hit the same issue with a very similar structure of directories you've mentioned in [this comment](https://github.com/sid88in/serverless-appsync-plugin/issues/194#issuecomment-486048544).

Have you somehow fixed / found a workaround on this?

For now I'm thinking of:
- using Serverless with JS/TS instead of YAML (if the `serverless-appsync-plugin` supports it...) to allow me to package and reuse templates as well as use the actual programming language to tinkle with sls templates in a sane way
- changing the deployment order (going  `1. top level serverless.yml with resources and graphql` -> `2. service abc` to `1. top level serverless.js with resources` -> `2. service abc` -> `3. serverless.js with appsync`)
- merging all the `schema.graphql` from my "microservices" into one using [this tool](https://www.graphql-tools.com/docs/schema-merging) 
- merging the `mappingTemplates` from my "microservices" into one using ...not sure what yet, but the expectation is that I dynamically create something as follows (and then convert the `yaml` into `js`):
```
mappingTemplates:
#### this comes from service abc
  - type: Query
    field: exampleQueryField
    dataSource: exampleDataSource
    request: @import_vtl_file_from_package_abc
    response: @import_vtl_file_from_package_abc
#### this comes from service def
  - type: Query
    field: exampleQueryFieldTwo
    dataSource: exampleDataSourceTwo
    request: @import_vtl_file_from_package_def
    response: @import_vtl_file_from_package_def
```

**Comment by sid88in** [comment]
This is the long-running "one shared AppSync API across multiple services" discussion, and it
really splits in two:

1. **Attaching resolvers/data sources from several services to one existing API (by id):**
   not supported on the current release (v1's `apiId` was removed), but being revived in #647
   (draft). Note that even with `apiId`, an AppSync API has a *single* schema defined in one
   place — so `apiId` lets services attach resolvers/data sources to a shared API, but it
   doesn't merge each service's schema fragment across separate deploys.

2. **Each service owning a schema fragment, merged into one API (federation):** that's the
   piece most people here actually want, and it's an AppSync limitation — there's no native
   federation (aws-appsync-community#35). Within a single service we do support schema
   stitching (globs + merge), just not across separate deployments.

Consolidating the cluster here and linking #647 for the attach work. The federation half
stays blocked on AWS. Thanks @EwanValentine @andrejunges @kkesley @tomaszdudek7 for all the
detail over the years.
